topo/node/ceos/ceos.go

```diff
@@ -66,99 +79,154 @@ func New(nodeImpl *node.Impl) (node.Node, error) {
-// For example, mock transport can be set via options
-func (n *Node) SpawnCLIConn() error {
-	opts := []scrapliutil.Option{
-		scrapliopts.WithAuthBypass(),
+func (n *Node) Create(ctx context.Context) error {
+	if err := n.CreateConfig(ctx); err != nil {
+		return fmt.Errorf("node %s failed to create config-map %w", n.Name(), err)
 	}
-
-	// add options defined in test package
-	opts = append(opts, n.testOpts...)
-
-	opts = n.PatchCLIConnOpen("kubectl", []string{"Cli"}, opts)
-
-	var err error
-	n.cliConn, err = n.GetCLIConn(scrapliPlatformName, opts)
-
-	return err
+	if err := n.CreateCRD(ctx); err != nil {
+		return fmt.Errorf("node %s failed to create custom resource definition %w", n.Name(), err)
+	}
+	return nil
 }
 
-func (n *Node) GenerateSelfSigned(ctx context.Context) error {
-	selfSigned := n.Proto.GetConfig().GetCert().GetSelfSigned()
-	if selfSigned == nil {
-		log.Infof("%s - no cert config", n.Name())
-		return nil
+func (n *Node) CreateCRD(ctx context.Context) error {
+	log.Infof("Creating new CEosLabDevice CRD for node: %v", n.Name())
+	proto := n.GetProto()
+	config := proto.GetConfig()
+	device := &ceos.CEosLabDevice{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "ceoslab.arista.com/v1alpha1",
+			Kind:       "CEosLabDevice",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      n.Name(),
+			Namespace: n.GetNamespace(),
+			Labels: map[string]string{
+				"app":  n.Name(),
+				"topo": n.GetNamespace(),
+			},
+		},
+		Spec: ceos.CEosLabDeviceSpec{
+			EnvVar:             config.GetEnv(),
+			Image:              config.GetImage(),
+			InitContainerImage: config.GetInitImage(),
+			Args:               config.GetArgs(),
+			Resources:          proto.GetConstraints(),
+			NumInterfaces:      uint32(len(proto.GetInterfaces())),
+			Sleep:              config.GetSleep(),
+		},
+	}
+	for label, v := range proto.GetLabels() {
+		device.ObjectMeta.Labels[label] = v
+	}
+	for port, service := range proto.GetServices() {
+		if device.Spec.Services == nil {
+			device.Spec.Services = map[string]ceos.ServiceConfig{}
+		}
+		device.Spec.Services[service.Name] = ceos.ServiceConfig{
+			TCPPorts: []uint32{port},
+		}
+	}
+	if cert := config.GetCert(); cert != nil {
+		if ssCert := cert.GetSelfSigned(); ssCert != nil {
+			certConfig := ceos.CertConfig{
+				SelfSignedCerts: []ceos.SelfSignedCertConfig{{
+					CertName:   ssCert.CertName,
+					KeyName:    ssCert.KeyName,
+					KeySize:    ssCert.KeySize,
+					CommonName: ssCert.CommonName,
+				}},
+			}
+			device.Spec.CertConfig = certConfig
+		}
+	}
+	for k, v := range n.GetProto().GetInterfaces() {
+		if device.Spec.IntfMapping == nil {
+			device.Spec.IntfMapping = map[string]string{}
+		}
+		device.Spec.IntfMapping[k] = v.GetName()
 	}
-	log.Infof("%s - generating self signed certs", n.Name())
-	log.Infof("%s - waiting for pod to be running", n.Name())
+	// Post to k8s
+	r := ceos.CEosLabDevice{}
+	err := n.KubeClient.CoreV1().RESTClient().
```

Comment:
sorry I missed this in the first review, can you abstract these details into a clientset and then call client.Create here (same with Delete)? An example is our CRD for topology: https://github.com/openconfig/kne/blob/b09a14a6ff36eb0a9448b5b95bac5d5ec9611da1/api/clientset/v1beta1/topology.go#L132

Here is an example a few months ago when we also had a rest client: https://github.com/openconfig/kne/blob/a2a782b79881cc7735cd11214313ebf45e6ea779/api/clientset/v1beta1/topology.go#L148

This will make testing a lot easier using fake clientsets

Reply:
okay, I defined a clientset in the operator. I may get to a fake clientset later but we don't need one for this PR. k8s code generators work poorly with operator-sdk.

Reply:
Thanks for adding that.

Btw, here is what our fake for our topology clientset looked like before we switched completely from the rest client to the dynamic client (which made testing easier): https://github.com/openconfig/kne/blob/3ad72bc0701eaaa7236f04bfe81d6a3495a93161/api/clientset/v1beta1/fake/fake.go (if that helps)

It would be nice at some point to have unit tests where a ceos is attempted to be created when one already exists, or attempted to be deleted when it DNE at some point when the fake is created.